lib/SubscribeClient.php

```diff
@@ -166,85 +147,91 @@ private function applyUri($uri) {
     }
 
     /**
-     * @return Promise
+     * @return void
      */
     public function close() {
-        $promises = [];
-
-        foreach ($this->promisors as $promisorGroup) {
-            foreach ($promisorGroup as $promisor) {
-                $promises[] = $promisor->promise();
-            }
-        }
-
-        foreach ($this->patternPromisors as $promisorGroup) {
-            foreach ($promisorGroup as $promisor) {
-                $promises[] = $promisor->promise();
-            }
-        }
-
-        /** @var Promise $promise */
-        $promise = all($promises);
-
-        $promise->when(function () {
-            $this->connection->close();
-        });
-
-        $this->unsubscribe();
-        $this->pUnsubscribe();
-
-        return $promise;
+        $this->connection->close();
     }
 
     /**
      * @param string $channel
-     * @return Promise
+     * @return Subscription
      */
     public function subscribe($channel) {
         $promisor = new Deferred;
+        $subscription = new Subscription($promisor, function($promisor) use ($channel) {
+            $this->unloadPromisor($promisor, $channel);
+        });
+
+        $this->promisors[$channel][spl_object_hash($promisor)] = $promisor;
 
         $promise = $this->connection->send(["subscribe", $channel]);
         $promise->when(function ($error) use ($channel, $promisor) {
             if ($error) {
+                $this->unloadPromisor($promisor, $channel);
                 $promisor->fail($error);
-            } else {
-                $this->promisors[$channel][] = $promisor;
-                $promisor->promise()->when(function () use ($channel) {
-                    array_shift($this->promisors[$channel]);
-                });
             }
         });
 
-        return $promisor->promise();
+        return $subscription;
+    }
+
+    private function unloadPromisor(Promisor $promisor, $channel) {
+        $hash = spl_object_hash($promisor);
+
+        if (isset($this->promisors[$channel][$hash])) {
+            unset($this->promisors[$channel][$hash]);
+
+            $promisor->succeed();
+
+            if (empty($this->promisors[$channel])) {
+                $this->unsubscribe($channel);
+            }
+        }
     }
 
     /**
      * @param string $pattern
-     * @return Promise
+     * @return Subscription
      */
     public function pSubscribe($pattern) {
         $promisor = new Deferred;
+        $subscription = new Subscription($promisor, function($promisor) use ($pattern) {
+            $this->unloadPatternPromisor($promisor, $pattern);
+        });
+
+        $this->patternPromisors[$pattern][spl_object_hash($promisor)] = $promisor;
```

Comment:
Why don't you just use() the $promisor too and just pass the promise (and not the Promisor) to Subscription?